File: dsa_leetcode_beginner/0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.java

import java.util.Stack;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

public class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int maxDepth = 0;
        Stack<TreeNodeDepthPair> stack = new Stack<>();
        stack.push(new TreeNodeDepthPair(root, 1));

        while (!stack.isEmpty()) {
            TreeNodeDepthPair current = stack.pop();
            TreeNode node = current.node;
            int depth = current.depth;

            if (node != null) {
                maxDepth = Math.max(maxDepth, depth);
                if (node.left != null) {
                    stack.push(new TreeNodeDepthPair(node.left, depth + 1));
                }
                if (node.right != null) {
                    stack.push(new TreeNodeDepthPair(node.right, depth + 1));
                }
            }
        }

        return maxDepth;
    }

    private class TreeNodeDepthPair {
        TreeNode node;
        int depth;

        TreeNodeDepthPair(TreeNode node, int depth) {
            this.node = node;
            this.depth = depth;
        }
    }
}

File: dsa_leetcode_beginner/0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.js

class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

function maxDepth(root) {
    if (root === null) {
        return 0;
    }

    let maxDepth = 0;
    const stack = [{ node: root, depth: 1 }];

    while (stack.length > 0) {
        const { node, depth } = stack.pop();

        if (node) {
            maxDepth = Math.max(maxDepth, depth);
            if (node.left) {
                stack.push({ node: node.left, depth: depth + 1 });
            }
            if (node.right) {
                stack.push({ node: node.right, depth: depth + 1 });
            }
        }
    }

    return maxDepth;
}
